### PR TITLE
Add common preprocessing for each request in node manager.

### PR DESCRIPTION
**Pull request by jiangzihao2009**
## What do these changes do?
1. Add common preprocessing for each request in node manager
2. Three parts are included in preprocessing, logging, checking heartbeat and checking being killed.
3. Change submit task request to an asynchronous request.

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: src/ray/raylet/node_manager.h
Comment (zhijunfu) [inline src/ray/raylet/node_manager.h]: nit: it might be better to rephrase this like:

Preprocess request from raylet client. This takes care of common processing for each request, such as logging, checking heartbeat and whether worker is killed.

File: src/ray/raylet/node_manager.cc
Comment (zhijunfu) [inline src/ray/raylet/node_manager.cc]: why removing this?

File: src/ray/raylet/node_manager.cc
Comment (zhijunfu) [inline src/ray/raylet/node_manager.cc]: nit: maybe better to just use `RAY_LOG` without using a separate `string_stream`?

File: src/ray/raylet/node_manager.cc
Comment (zhijunfu) [inline src/ray/raylet/node_manager.cc]: Maybe simpler to use two functions `GetWorker` and `GetDriver` so that the semantics can be more clear?

File: src/ray/raylet/node_manager.cc
Comment (zhijunfu) [inline src/ray/raylet/node_manager.cc]: Could we just use RAY_LOG(DEBUG) for the debug messages, instead of `straing_stream`? Thus we don't need to allocate these debug strings on production (where debug log is disabled).

File: src/ray/raylet/node_manager.cc
Comment (zhijunfu) [inline src/ray/raylet/node_manager.cc]: do we need to check this for every request? Maybe reduce it to every 10 requests to reduce the time calculation cost?

**Review by zhijunfu — COMMENTED** [review]
Thanks! I left a few comments.

File: src/ray/raylet/node_manager.cc
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: Better also add `request` as a parameter of this macro. Because otherwise, it's unclear what this `request` is.

File: src/ray/raylet/node_manager.cc
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: The `request` parameter isn't being used?

File: src/ray/raylet/node_manager.cc
Comment (raulchen) [reply src/ray/raylet/node_manager.cc]: I think it should be very cheap to calculate this?

File: src/ray/raylet/node_manager.cc
Comment (jiangzihao2009) [reply src/ray/raylet/node_manager.cc]: Not used, a worker can be a worker or a driver.

**Comment by raulchen** [comment]
Also, can you explain that the main purpose of this PR is to solve the issue that `the heartbeat timer will be delayed and makes a node mistakenly marked as dead, when node manager suddenly receives many requests and posts many handlers to the event loop`?

File: src/ray/raylet/node_manager.cc
Comment (jiangzihao2009) [reply src/ray/raylet/node_manager.cc]: I think we need a function to get the pointer to the worker according to the worker id. We still need to check the return pointer to check whether it's a driver or worker if use two funcs.

Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: yes. An issue with original code is the second field `is_worker` in the returned pair is meaningless when the first field is nullptr.This is a minor issue though.

File: src/ray/raylet/node_manager.cc
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: Theoretically we should add the check for all the messages in raylet? node manager, object manager, and from worker to raylet.

Comment (raulchen) [reply src/ray/raylet/node_manager.cc]: Yes, but except object manager requests, which are handled in background threads.

File: src/ray/raylet/node_manager.cc
Comment (jiangzihao2009) [reply src/ray/raylet/node_manager.cc]: You mean just return here and move the log in PreprocessRequest?

Comment (raulchen) [reply src/ray/raylet/node_manager.cc]: yes

File: src/ray/protobuf/raylet.proto
Comment (raulchen) [inline src/ray/protobuf/raylet.proto]:
```suggestion
/// NOTE(Joey Jiang) Every request defined in this file should have a `worker_id` field, which
```

File: src/ray/protobuf/raylet.proto
Comment (raulchen) [inline src/ray/protobuf/raylet.proto]:
```suggestion
/// will be used in `NodeManager::PreprocessRequest`.
```

File: src/ray/raylet/node_manager.cc
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: This warning is duplicated with the one in `Heartbeat`. I think we can just delete it.

File: src/ray/raylet/node_manager.cc
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: I tried some stressful tests, 50% timeout still can't trigger the timer. maybe we can be aggressive and send heartbeat if delay > one heartbeat interval.

File: src/ray/raylet/node_manager.cc
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]:
```suggestion
  // NOTE(Joey Jiang): We check heartbeat proactively because the heartbeats' callback won't
```

File: src/ray/raylet/node_manager.cc
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: print request type here?

File: src/ray/raylet/node_manager.cc
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: same here

File: src/ray/raylet/node_manager.cc
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: instead of returning a pair here, I think it's clearer to put `is_worker` in `Worker` class.

**Review by raulchen — COMMENTED** [review]
Looks good overall. Just a few small comments.

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: It would be great if this could also print the type of the request. Also, you can just write `RAY_LOG(DEBUG) << ...`, instead of `oss << ...; RAY_LOG(DEBUG) << oss.str()`.

File: src/ray/raylet/node_manager.h
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.h]: Be specific about each of these steps. Logging is self-explanatory, but for instance, I cannot tell from this comment what "checking heartbeat" means, or what will happen if the "worker is being killed".

File: src/ray/raylet/node_manager.h
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.h]: This param does not seem to exist.

File: src/ray/raylet/node_manager.h
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.h]: Templates are not params.

File: src/ray/raylet/node_manager.h
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.h]: Document the return value and what it means if true/false is returned.

**Review by stephanie-wang — CHANGES_REQUESTED** [review]
This seems like a pretty brittle and at-best temporary solution, but I guess it's okay for now. How feasible would it be to move the heartbeat handling to a separate thread? Is it necessary to use heartbeats for worker-raylet communication?

**Comment by pcmoritz** [comment]
I also think moving the heartbeats to a different thread is the right solution here. We do need to make sure that it is not possible for the main thread to get stuck and the heartbeat thread to keep sending heartbeats however.

File: src/ray/raylet/node_manager.cc
Comment (jiangzihao2009) [reply src/ray/raylet/node_manager.cc]: The basic idea here is to send the heartbeat proactively to the raylet monitor, which prevents the timeout of this node. The root cause of this problem is that too many requests posted in the main thread in a short time. I have thought about using a seperate thread to send heartbeat, we might need a extra gcs client and some locks, we will talk about how to do it better.

File: src/ray/raylet/node_manager.cc
Comment (jiangzihao2009) [reply src/ray/raylet/node_manager.cc]: I need to print only one DEBUG log for a request.

**Comment by raulchen** [comment]
@stephanie-wang @pcmoritz The problem with moving heartbeat to a different thread is that we need to use the gcs client to send heartbeat. However, the gcs client is bound to the io_service in the main thread.

File: src/ray/raylet/node_manager.cc
Comment (raulchen) [reply src/ray/raylet/node_manager.cc]: @stephanie-wang because if raylet suddenly receives a huge number of `ForwardTask` requests,  raylet-to-gcs heartbeat will time out as well.

**Comment by raulchen** [comment]
I think the actual problem is that asio doesn't prioritize timers over post handlers (AFAIK, some other event loops, e.g., Python's event loop, do.) Otherwise, it's totally fine to put timer in the event loop.

**Comment by raulchen** [comment]
@pcmoritz @stephanie-wang Since it's not easy to move heartbeat timer to another thread, let's use this PR to unblock the issue for now?

File: src/ray/raylet/node_manager.h
Comment (raulchen) [inline src/ray/raylet/node_manager.h]: typo `exiting`

**Review by raulchen — APPROVED** [review]
LGTM. just one typo

**Comment by simon-mo** [comment]
I wonder how much fixed overhead does this preprocessing task will create? Can we quickly check how much slower does it create?

**Comment by raulchen** [comment]
@simon-mo checking timer should be very cheap. my local test with your script shows it's even slightly faster than master.

**Comment by stephanie-wang** [comment]
After some discussion with @pcmoritz and @simon-mo, let's not merge this PR unless it is really necessary for stability even after #5341 has been merged. For raylet-to-redis heartbeats, the issue has been known for a while and the solution needs to be more comprehensive than this PR. For worker heartbeats, we think the long-term solution is to remove them completely, since it is possible for the raylet to detect worker failures without heartbeats. We're concerned that merging this code will encourage further short-term fixes when really we should concentrate on fixing the core issues.

**Comment by jiangzihao2009** [comment]
> After some discussion with @pcmoritz and @simon-mo, let's not merge this PR unless it is really necessary for stability even after #5341 has been merged. For raylet-to-redis heartbeats, the issue has been known for a while and the solution needs to be more comprehensive than this PR. For worker heartbeats, we think the long-term solution is to remove them completely, since it is possible for the raylet to detect worker failures without heartbeats. We're concerned that merging this code will encourage further short-term fixes when really we should concentrate on fixing the core issues.

At first, this PR is only intended to print log and check whether the job is stopped, which are missing part of PR #5120. We found the heartbeat problem in some test and tried to solve it here. In fact, the root cause can be concluded as a misusing of GRPC. And I think #5341 is a better solution to solve the heartbeat problem.

**Comment by raulchen** [comment]
I think we can revert the heartbeat change, and merge other changes in this PR.

**Comment by raulchen** [comment]
@stephanie-wang heartbeat timer is reverted. I think this PR can be merged. Please let me know if you have other comments.

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Move this inside the `do {}` block so it is not in scope after the macro.

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Can we name this and the macro something more descriptive of what the macro does? Something like `RETURN_IF_WORKER_DEAD` and `CheckWorkerDead`?

**Review by stephanie-wang — APPROVED** [review]
Thanks!

File: src/ray/raylet/node_manager.cc
Comment (raulchen) [reply src/ray/raylet/node_manager.cc]: This macro does some common preprocessing work for every worker requests. Currently it does 2 things: logging request and checking worker dead. And we may need it to do more things in the future. So `CheckWorkerDead` doesn't describe the whole purposes. 
I think we can name this `PREPROCESS_WORKER_REQUEST` to be more specific.

**Pull request merged**
